File: fabric/src/main/java/wile/redstonepen/gametest/ControlBoxGameTests.java
package wile.redstonepen.gametest;

import net.minecraft.gametest.framework.GameTest;
import net.minecraft.gametest.framework.GameTestHelper;
import wile.redstonepen.gametestcommon.ControlBoxTests;

public class ControlBoxGameTests
{
  private static final String TEMPLATE = "redstonepen:relay_activates_from_redstone";

  public ControlBoxGameTests() {}

  @GameTest(template = TEMPLATE, timeoutTicks = 20)
  public static void controlBoxEvaluatesConstantProgram(GameTestHelper helper)
  { ControlBoxTests.controlBoxEvaluatesConstantProgram(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 20)
  public static void controlBoxRejectsInvalidProgram(GameTestHelper helper)
  { ControlBoxTests.controlBoxRejectsInvalidProgram(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 5)
  public static void writenbtRoundTripsCodeAndSymbols(GameTestHelper helper)
  { ControlBoxTests.writenbtRoundTripsCodeAndSymbols(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 5)
  public static void readnbtRestoresCodeAndOutputData(GameTestHelper helper)
  { ControlBoxTests.readnbtRestoresCodeAndOutputData(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 5)
  public static void readnbtPreservesSymbolMap(GameTestHelper helper)
  { ControlBoxTests.readnbtPreservesSymbolMap(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 5)
  public static void initiallyDisabled(GameTestHelper helper)
  { ControlBoxTests.initiallyDisabled(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 5)
  public static void setEnabledTrueFlipsState(GameTestHelper helper)
  { ControlBoxTests.setEnabledTrueFlipsState(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 5)
  public static void setEnabledIdempotent(GameTestHelper helper)
  { ControlBoxTests.setEnabledIdempotent(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 10)
  public static void tickOnDisabledBoxClearsOutput(GameTestHelper helper)
  { ControlBoxTests.tickOnDisabledBoxClearsOutput(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 10)
  public static void tickWithInvalidCodeDoesNotThrow(GameTestHelper helper)
  { ControlBoxTests.tickWithInvalidCodeDoesNotThrow(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 5)
  public static void traceToggleFlipsFlag(GameTestHelper helper)
  { ControlBoxTests.traceToggleFlipsFlag(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 5)
  public static void getNameFallsBackToBlockTranslationKey(GameTestHelper helper)
  { ControlBoxTests.getNameFallsBackToBlockTranslationKey(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 5)
  public static void setCustomNameStoresAndReturnsIt(GameTestHelper helper)
  { ControlBoxTests.setCustomNameStoresAndReturnsIt(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 5)
  public static void onServerPacketReceivedAppliesCodeFromNbt(GameTestHelper helper)
  { ControlBoxTests.onServerPacketReceivedAppliesCodeFromNbt(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 10)
  public static void getSignalReturnsComputedOutput(GameTestHelper helper)
  { ControlBoxTests.getSignalReturnsComputedOutput(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 10)
  public static void getDirectSignalMatchesGetSignal(GameTestHelper helper)
  { ControlBoxTests.getDirectSignalMatchesGetSignal(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 5)
  public static void dropListWithCodeSavesNbt(GameTestHelper helper)
  { ControlBoxTests.dropListWithCodeSavesNbt(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 5)
  public static void dropListWithNoCodeHasNoNbt(GameTestHelper helper)
  { ControlBoxTests.dropListWithNoCodeHasNoNbt(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 5)
  public static void setPlacedByWithNbtRestoresCode(GameTestHelper helper)
  { ControlBoxTests.setPlacedByWithNbtRestoresCode(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 5)
  public static void setPlacedByWithEmptyNbtIsNoOp(GameTestHelper helper)
  { ControlBoxTests.setPlacedByWithEmptyNbtIsNoOp(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 5)
  public static void updateWithNullFromPosResetsTickTimer(GameTestHelper helper)
  { ControlBoxTests.updateWithNullFromPosResetsTickTimer(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 5)
  public static void updateWithNeighborPosTriggersSideScan(GameTestHelper helper)
  { ControlBoxTests.updateWithNeighborPosTriggersSideScan(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 5)
  public static void isBlockEntityTickingAlwaysTrue(GameTestHelper helper)
  { ControlBoxTests.isBlockEntityTickingAlwaysTrue(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 5)
  public static void defsDefaultConstructorIsCallable(GameTestHelper helper)
  { ControlBoxTests.defsDefaultConstructorIsCallable(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 5)
  public static void getDisplayNameReturnsNonNull(GameTestHelper helper)
  { ControlBoxTests.getDisplayNameReturnsNonNull(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 5)
  public static void createMenuReturnsNonNull(GameTestHelper helper)
  { ControlBoxTests.createMenuReturnsNonNull(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 10)
  public static void tickWithTraceEnabledCoversTracePaths(GameTestHelper helper)
  { ControlBoxTests.tickWithTraceEnabledCoversTracePaths(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 10)
  public static void tickWithTickrateSymbolSetsInterval(GameTestHelper helper)
  { ControlBoxTests.tickWithTickrateSymbolSetsInterval(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 5)
  public static void setEnabledFalseFromEnabledClearsSymbols(GameTestHelper helper)
  { ControlBoxTests.setEnabledFalseFromEnabledClearsSymbols(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 5)
  public static void setRcaPlayerUuidNonNullDoesNotThrow(GameTestHelper helper)
  { ControlBoxTests.setRcaPlayerUuidNonNullDoesNotThrow(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 10)
  public static void signalUpdateRisingEdgeSetsIntredge(GameTestHelper helper)
  { ControlBoxTests.signalUpdateRisingEdgeSetsIntredge(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 10)
  public static void signalUpdateFallingEdgeSetsIntfedge(GameTestHelper helper)
  { ControlBoxTests.signalUpdateFallingEdgeSetsIntfedge(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 5)
  public static void useItemOnWithDebugStickTogglesTrace(GameTestHelper helper)
  { ControlBoxTests.useItemOnWithDebugStickTogglesTrace(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 10)
  public static void controlBoxBuiltinFunctionsExercised(GameTestHelper helper)
  { ControlBoxTests.controlBoxBuiltinFunctionsExercised(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 10)
  public static void controlBoxCounterFunctionsExercised(GameTestHelper helper)
  { ControlBoxTests.controlBoxCounterFunctionsExercised(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 10)
  public static void controlBoxTimerFunctionsExercised(GameTestHelper helper)
  { ControlBoxTests.controlBoxTimerFunctionsExercised(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 10)
  public static void controlBoxRemainingTimerVariantsExercised(GameTestHelper helper)
  { ControlBoxTests.controlBoxRemainingTimerVariantsExercised(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 10)
  public static void controlBoxTimerEdgeCasesDoNotThrow(GameTestHelper helper)
  { ControlBoxTests.controlBoxTimerEdgeCasesDoNotThrow(helper); }

  @GameTest(template = TEMPLATE, timeoutTicks = 5)
  public static void controlBoxSetCodeSameCodeSkipsReparse(GameTestHelper helper)
  { ControlBoxTests.controlBoxSetCodeSameCodeSkipsReparse(helper); }
}

File: neoforge/src/main/kotlin/wile/redstonepen/gametest/ControlBoxGameTests.kt
package wile.redstonepen.gametest

import net.minecraft.gametest.framework.GameTest
import net.minecraft.gametest.framework.GameTestHelper
import net.neoforged.neoforge.gametest.GameTestHolder
import net.neoforged.neoforge.gametest.PrefixGameTestTemplate
import wile.redstonepen.ModConstants
import wile.redstonepen.gametestcommon.ControlBoxTests

@GameTestHolder(ModConstants.MODID)
@PrefixGameTestTemplate(false)
object ControlBoxGameTests {
    private const val TEMPLATE = "relay_activates_from_redstone"

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 20)
    fun controlBoxEvaluatesConstantProgram(helper: GameTestHelper) =
        ControlBoxTests.controlBoxEvaluatesConstantProgram(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 20)
    fun controlBoxRejectsInvalidProgram(helper: GameTestHelper) =
        ControlBoxTests.controlBoxRejectsInvalidProgram(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 5)
    fun writenbtRoundTripsCodeAndSymbols(helper: GameTestHelper) =
        ControlBoxTests.writenbtRoundTripsCodeAndSymbols(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 5)
    fun readnbtRestoresCodeAndOutputData(helper: GameTestHelper) =
        ControlBoxTests.readnbtRestoresCodeAndOutputData(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 5)
    fun readnbtPreservesSymbolMap(helper: GameTestHelper) =
        ControlBoxTests.readnbtPreservesSymbolMap(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 5)
    fun initiallyDisabled(helper: GameTestHelper) = ControlBoxTests.initiallyDisabled(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 5)
    fun setEnabledTrueFlipsState(helper: GameTestHelper) =
        ControlBoxTests.setEnabledTrueFlipsState(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 5)
    fun setEnabledIdempotent(helper: GameTestHelper) = ControlBoxTests.setEnabledIdempotent(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 10)
    fun tickOnDisabledBoxClearsOutput(helper: GameTestHelper) =
        ControlBoxTests.tickOnDisabledBoxClearsOutput(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 10)
    fun tickWithInvalidCodeDoesNotThrow(helper: GameTestHelper) =
        ControlBoxTests.tickWithInvalidCodeDoesNotThrow(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 5)
    fun traceToggleFlipsFlag(helper: GameTestHelper) = ControlBoxTests.traceToggleFlipsFlag(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 5)
    fun getNameFallsBackToBlockTranslationKey(helper: GameTestHelper) =
        ControlBoxTests.getNameFallsBackToBlockTranslationKey(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 5)
    fun setCustomNameStoresAndReturnsIt(helper: GameTestHelper) =
        ControlBoxTests.setCustomNameStoresAndReturnsIt(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 5)
    fun onServerPacketReceivedAppliesCodeFromNbt(helper: GameTestHelper) =
        ControlBoxTests.onServerPacketReceivedAppliesCodeFromNbt(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 10)
    fun getSignalReturnsComputedOutput(helper: GameTestHelper) =
        ControlBoxTests.getSignalReturnsComputedOutput(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 10)
    fun getDirectSignalMatchesGetSignal(helper: GameTestHelper) =
        ControlBoxTests.getDirectSignalMatchesGetSignal(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 5)
    fun dropListWithCodeSavesNbt(helper: GameTestHelper) =
        ControlBoxTests.dropListWithCodeSavesNbt(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 5)
    fun dropListWithNoCodeHasNoNbt(helper: GameTestHelper) =
        ControlBoxTests.dropListWithNoCodeHasNoNbt(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 5)
    fun setPlacedByWithNbtRestoresCode(helper: GameTestHelper) =
        ControlBoxTests.setPlacedByWithNbtRestoresCode(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 5)
    fun setPlacedByWithEmptyNbtIsNoOp(helper: GameTestHelper) =
        ControlBoxTests.setPlacedByWithEmptyNbtIsNoOp(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 5)
    fun updateWithNullFromPosResetsTickTimer(helper: GameTestHelper) =
        ControlBoxTests.updateWithNullFromPosResetsTickTimer(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 5)
    fun updateWithNeighborPosTriggersSideScan(helper: GameTestHelper) =
        ControlBoxTests.updateWithNeighborPosTriggersSideScan(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 5)
    fun isBlockEntityTickingAlwaysTrue(helper: GameTestHelper) =
        ControlBoxTests.isBlockEntityTickingAlwaysTrue(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 5)
    fun defsPortNamesHasSixEntries(helper: GameTestHelper) =
        ControlBoxTests.defsPortNamesHasSixEntries(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 5)
    fun getDisplayNameReturnsNonNull(helper: GameTestHelper) =
        ControlBoxTests.getDisplayNameReturnsNonNull(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 5)
    fun createMenuReturnsNonNull(helper: GameTestHelper) =
        ControlBoxTests.createMenuReturnsNonNull(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 10)
    fun tickWithTraceEnabledCoversTracePaths(helper: GameTestHelper) =
        ControlBoxTests.tickWithTraceEnabledCoversTracePaths(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 10)
    fun tickWithTickrateSymbolSetsInterval(helper: GameTestHelper) =
        ControlBoxTests.tickWithTickrateSymbolSetsInterval(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 5)
    fun setEnabledFalseFromEnabledClearsSymbols(helper: GameTestHelper) =
        ControlBoxTests.setEnabledFalseFromEnabledClearsSymbols(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 5)
    fun setRcaPlayerUuidNonNullDoesNotThrow(helper: GameTestHelper) =
        ControlBoxTests.setRcaPlayerUuidNonNullDoesNotThrow(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 10)
    fun signalUpdateRisingEdgeSetsIntredge(helper: GameTestHelper) =
        ControlBoxTests.signalUpdateRisingEdgeSetsIntredge(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 10)
    fun signalUpdateFallingEdgeSetsIntfedge(helper: GameTestHelper) =
        ControlBoxTests.signalUpdateFallingEdgeSetsIntfedge(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 5)
    fun useItemOnWithDebugStickTogglesTrace(helper: GameTestHelper) =
        ControlBoxTests.useItemOnWithDebugStickTogglesTrace(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 10)
    fun controlBoxBuiltinFunctionsExercised(helper: GameTestHelper) =
        ControlBoxTests.controlBoxBuiltinFunctionsExercised(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 10)
    fun controlBoxCounterFunctionsExercised(helper: GameTestHelper) =
        ControlBoxTests.controlBoxCounterFunctionsExercised(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 10)
    fun controlBoxTimerFunctionsExercised(helper: GameTestHelper) =
        ControlBoxTests.controlBoxTimerFunctionsExercised(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 10)
    fun controlBoxRemainingTimerVariantsExercised(helper: GameTestHelper) =
        ControlBoxTests.controlBoxRemainingTimerVariantsExercised(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 10)
    fun controlBoxTimerEdgeCasesDoNotThrow(helper: GameTestHelper) =
        ControlBoxTests.controlBoxTimerEdgeCasesDoNotThrow(helper)

    @JvmStatic
    @GameTest(template = TEMPLATE, timeoutTicks = 5)
    fun controlBoxSetCodeSameCodeSkipsReparse(helper: GameTestHelper) =
        ControlBoxTests.controlBoxSetCodeSameCodeSkipsReparse(helper)
}
